File: backend-node/services/statementGenerator.js
const { CHART } = require('./chartOfAccounts');

function fmt(n) { return Math.round(n * 100) / 100; }

function generateStatements(mappings, tbRecords, meta) {
  // Build lookup: account_code -> balance
  const balanceMap = {};
  for (const r of tbRecords) {
    balanceMap[r.account_code] = r.balance;
    if (!balanceMap['__name__' + r.account_name]) balanceMap['__name__' + r.account_name] = r.balance;
  }

  // Group by statement -> category -> list of {label, amount}
  const grouped = {};
  for (const m of mappings) {
    if (m.statement === 'unmapped') continue;
    const stmt = m.statement || 'balance_sheet';
    const cat  = m.category  || 'Current Assets';
    if (!grouped[stmt]) grouped[stmt] = {};
    if (!grouped[stmt][cat]) grouped[stmt][cat] = [];

    const balance = balanceMap[m.account_code] ?? balanceMap['__name__' + m.account_name] ?? 0;
    const sign    = m.sign || 1;
    grouped[stmt][cat].push({ label: m.account_name, amount: fmt(balance * sign) });
  }

  // Unmapped accounts
  const unmapped = mappings.filter(m => m.statement === 'unmapped');

  // Helper: build a section dict for one statement
  function buildSections(stmt, categoryOrder) {
    const sections = {};
    const available = grouped[stmt] || {};
    // Put in preferred order first, then any extras
    const allCats = [...new Set([...categoryOrder, ...Object.keys(available)])];
    for (const cat of allCats) {
      const items = available[cat];
      if (!items || !items.length) continue;
      sections[cat] = {
        items,
        total: fmt(items.reduce((a, b) => a + b.amount, 0)),
      };
    }
    return sections;
  }

  // ── Balance Sheet ───────────────────────────────────────────
  const bsSections = buildSections('balance_sheet', [
    'Current Assets','Non-Current Assets',
    'Current Liabilities','Non-Current Liabilities','Equity',
  ]);

  let totalAssets = 0, totalLiabilities = 0, totalEquity = 0;
  for (const [cat, sec] of Object.entries(bsSections)) {
    if (['Current Assets','Non-Current Assets'].includes(cat))         totalAssets      += sec.total;
    else if (['Current Liabilities','Non-Current Liabilities'].includes(cat)) totalLiabilities += sec.total;
    else if (cat === 'Equity')                                          totalEquity      += sec.total;
  }
  totalAssets      = fmt(totalAssets);
  totalLiabilities = fmt(totalLiabilities);
  totalEquity      = fmt(totalEquity);

  const balance_sheet = {
    sections: bsSections,
    total_assets:               totalAssets,
    total_liabilities:          totalLiabilities,
    total_equity:               totalEquity,
    total_liabilities_and_equity: fmt(totalLiabilities + totalEquity),
    balance_check:              fmt(totalAssets - (totalLiabilities + totalEquity)),
  };

  // ── Income Statement ────────────────────────────────────────
  const isSections = buildSections('income_statement', [
    'Revenue','Cost of Sales','Operating Expenses','Finance Costs','Income Tax',
  ]);

  const revenue  = isSections['Revenue']?.total             || 0;
  const cogs     = isSections['Cost of Sales']?.total       || 0;
  const opex     = isSections['Operating Expenses']?.total  || 0;
  const finCost  = isSections['Finance Costs']?.total       || 0;
  const tax      = isSections['Income Tax']?.total          || 0;

  const grossProfit = fmt(revenue - cogs);
  const ebit        = fmt(grossProfit - opex);
  const ebt         = fmt(ebit - finCost);
  const netIncome   = fmt(ebt - tax);

  const income_statement = {
    sections: isSections,
    revenue, gross_profit: grossProfit, ebit, ebt, net_income: netIncome,
  };

  // ── Cash Flow ───────────────────────────────────────────────
  const cfSections = buildSections('cash_flow', [
    'Operating Activities','Investing Activities','Financing Activities',
  ]);

  let operating = cfSections['Operating Activities']?.total || 0;
  let investing = cfSections['Investing Activities']?.total || 0;
  let financing = cfSections['Financing Activities']?.total || 0;

  // If no CF data mapped, derive operating from net income (indirect method)
  if (!Object.keys(cfSections).length) {
    cfSections['Operating Activities'] = {
      items: [{ label: 'Net Income (Indirect Method)', amount: netIncome }],
      total: netIncome,
    };
    operating = netIncome;
    cfSections['Investing Activities']  = { items: [], total: 0 };
    cfSections['Financing Activities']  = { items: [], total: 0 };
  }

  const cash_flow = {
    sections: cfSections,
    net_change_in_cash: fmt(operating + investing + financing),
  };

  // ── Equity Statement ────────────────────────────────────────
  const openingEquity = fmt(totalEquity - netIncome);
  const equity_statement = {
    opening_equity:  openingEquity,
    net_income:      netIncome,
    dividends:       0,
    other_movements: 0,
    closing_equity:  totalEquity,
  };

  // ── KPIs ────────────────────────────────────────────────────
  const kpis = {
    total_assets:      totalAssets,
    total_liabilities: totalLiabilities,
    total_equity:      totalEquity,
    revenue:           fmt(revenue),
    gross_profit:      fmt(grossProfit),
    net_income:        fmt(netIncome),
    net_cash_flow:     fmt(operating + investing + financing),
    gross_margin:      revenue ? fmt((grossProfit / revenue) * 100) : 0,
    net_margin:        revenue ? fmt((netIncome   / revenue) * 100) : 0,
  };

  return {
    balance_sheet,
    income_statement,
    cash_flow,
    equity_statement,
    kpis,
    unmapped,
    entity_name: meta?.entity_name || '',
    period_end:  meta?.period_end  || '',
    currency:    meta?.currency    || 'USD',
  };
}

module.exports = { generateStatements };

File: backend/services/statement_generator.py
from typing import List, Dict, Any
from models.schemas import AccountMapping
from models.chart_of_accounts import CHART_OF_ACCOUNTS


def generate_statements(
    trial_balance: List[Dict[str, Any]],
    mappings: List[AccountMapping],
    entity_name: str,
    period_end: str,
    currency: str = "USD",
) -> Dict[str, Any]:
    mapping_index = {m.account_code or m.account_name: m for m in mappings}

    buckets: Dict[str, Dict[str, Dict[str, float]]] = {
        "balance_sheet": {},
        "income_statement": {},
        "cash_flow": {},
        "equity": {},
        "unmapped": {"Unmapped": {"Unmapped": 0.0}},
    }

    line_items: Dict[str, List[Dict]] = {
        "balance_sheet": [],
        "income_statement": [],
        "cash_flow": [],
        "equity": [],
        "unmapped": [],
    }

    for entry in trial_balance:
        key = entry.get("account_code") or entry.get("account_name")
        mapping = mapping_index.get(key) or _find_by_name(mapping_index, entry.get("account_name", ""))
        balance = entry.get("balance", 0.0)

        if mapping and mapping.statement != "unmapped":
            stmt = mapping.statement
            cat = mapping.category
            sub = mapping.subcategory
            signed_balance = balance * mapping.sign
        else:
            stmt = "unmapped"
            cat = "Unmapped"
            sub = "Unmapped"
            signed_balance = balance

        if cat not in buckets.get(stmt, {}):
            buckets.setdefault(stmt, {})[cat] = {}
        if sub not in buckets[stmt][cat]:
            buckets[stmt][cat][sub] = 0.0
        buckets[stmt][cat][sub] += signed_balance

        line_items.setdefault(stmt, []).append({
            "account_code": entry.get("account_code", ""),
            "account_name": entry.get("account_name", ""),
            "category": cat,
            "subcategory": sub,
            "balance": round(signed_balance, 2),
        })

    balance_sheet = _build_balance_sheet(buckets.get("balance_sheet", {}))
    income_statement = _build_income_statement(buckets.get("income_statement", {}))
    cash_flow = _build_cash_flow(buckets.get("cash_flow", {}), income_statement)
    equity_statement = _build_equity_statement(buckets.get("equity", {}), income_statement)

    return {
        "entity_name": entity_name,
        "period_end": period_end,
        "currency": currency,
        "balance_sheet": balance_sheet,
        "income_statement": income_statement,
        "cash_flow": cash_flow,
        "equity_statement": equity_statement,
        "unmapped": line_items.get("unmapped", []),
        "line_items": line_items,
    }


def _find_by_name(mapping_index: Dict, name: str):
    return mapping_index.get(name)


def _build_balance_sheet(buckets: Dict) -> Dict:
    sections = {}
    for category, subcats in buckets.items():
        items = [{"label": k, "amount": round(v, 2)} for k, v in subcats.items()]
        total = round(sum(v for v in subcats.values()), 2)
        sections[category] = {"items": items, "total": total}

    total_assets = round(
        sum(sections.get(c, {}).get("total", 0) for c in ["Current Assets", "Non-Current Assets"]),
        2,
    )
    total_current_liab = sections.get("Current Liabilities", {}).get("total", 0)
    total_noncurrent_liab = sections.get("Non-Current Liabilities", {}).get("total", 0)
    total_liabilities = round(total_current_liab + total_noncurrent_liab, 2)
    total_equity = sections.get("Equity", {}).get("total", 0)
    total_liab_equity = round(total_liabilities + total_equity, 2)
    balance_check = round(total_assets - total_liab_equity, 2)

    return {
        "sections": sections,
        "total_assets": total_assets,
        "total_current_liabilities": round(total_current_liab, 2),
        "total_non_current_liabilities": round(total_noncurrent_liab, 2),
        "total_liabilities": total_liabilities,
        "total_equity": round(total_equity, 2),
        "total_liabilities_and_equity": total_liab_equity,
        "balance_check": balance_check,
    }


def _build_income_statement(buckets: Dict) -> Dict:
    sections = {}
    for category, subcats in buckets.items():
        items = [{"label": k, "amount": round(v, 2)} for k, v in subcats.items()]
        total = round(sum(v for v in subcats.values()), 2)
        sections[category] = {"items": items, "total": total}

    revenue = sections.get("Revenue", {}).get("total", 0)
    cogs = sections.get("Cost of Sales", {}).get("total", 0)
    gross_profit = round(revenue - cogs, 2)

    opex = sections.get("Operating Expenses", {}).get("total", 0)
    ebit = round(gross_profit - opex, 2)

    finance_costs = sections.get("Finance Costs", {}).get("total", 0)
    ebt = round(ebit - finance_costs, 2)

    tax = sections.get("Income Tax", {}).get("total", 0)
    net_income = round(ebt - tax, 2)

    other_income_items = sections.get("Revenue", {}).get("items", [])
    other_income = sum(i["amount"] for i in other_income_items if "other" in i["label"].lower())

    return {
        "sections": sections,
        "revenue": round(revenue, 2),
        "cost_of_sales": round(cogs, 2),
        "gross_profit": gross_profit,
        "operating_expenses": round(opex, 2),
        "ebit": ebit,
        "finance_costs": round(finance_costs, 2),
        "ebt": ebt,
        "income_tax": round(tax, 2),
        "net_income": net_income,
    }


def _build_cash_flow(buckets: Dict, income_stmt: Dict) -> Dict:
    sections = {}
    for category, subcats in buckets.items():
        items = [{"label": k, "amount": round(v, 2)} for k, v in subcats.items()]
        total = round(sum(v for v in subcats.values()), 2)
        sections[category] = {"items": items, "total": total}

    # If no cash flow data uploaded, build indirect method shell from P&L
    if not sections:
        net_income = income_stmt.get("net_income", 0)
        sections["Operating Activities"] = {
            "items": [
                {"label": "Net Income", "amount": net_income},
                {"label": "Adjustments (enter manually)", "amount": 0},
            ],
            "total": net_income,
        }
        sections["Investing Activities"] = {"items": [], "total": 0}
        sections["Financing Activities"] = {"items": [], "total": 0}

    operating = sections.get("Operating Activities", {}).get("total", 0)
    investing = sections.get("Investing Activities", {}).get("total", 0)
    financing = sections.get("Financing Activities", {}).get("total", 0)
    net_change = round(operating + investing + financing, 2)

    return {
        "sections": sections,
        "net_cash_operating": round(operating, 2),
        "net_cash_investing": round(investing, 2),
        "net_cash_financing": round(financing, 2),
        "net_change_in_cash": net_change,
    }


def _build_equity_statement(buckets: Dict, income_stmt: Dict) -> Dict:
    net_income = income_stmt.get("net_income", 0)
    sections = {}
    for category, subcats in buckets.items():
        items = [{"label": k, "amount": round(v, 2)} for k, v in subcats.items()]
        total = round(sum(v for v in subcats.values()), 2)
        sections[category] = {"items": items, "total": total}

    changes = sections.get("Changes in Equity", {}).get("items", [])
    opening = next((i["amount"] for i in changes if "opening" in i["label"].lower()), 0)
    dividends = next((i["amount"] for i in changes if "dividend" in i["label"].lower()), 0)
    other = sum(i["amount"] for i in changes if "opening" not in i["label"].lower() and "dividend" not in i["label"].lower() and "net profit" not in i["label"].lower())
    closing = round(opening + net_income - dividends + other, 2)

    return {
        "sections": sections,
        "opening_equity": round(opening, 2),
        "net_income": round(net_income, 2),
        "dividends": round(dividends, 2),
        "other_movements": round(other, 2),
        "closing_equity": closing,
    }
